akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.stream.ActorMaterializer
+import akka.http.scaladsl.settings.ServerSettings
+import com.typesafe.config.{ Config, ConfigFactory }
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, ExecutionContext, Future }
+import scala.io.StdIn
+import scala.util.{ Failure, Success }
+
+/**
+ * Bootstrap trait. It helps booting up an akka-http by only defining the desired routes.
+ * It offers additional hooks to modify the default behavior
+ */
+trait HttpApp {
+
+  def main(args: Array[String]) = {
+    val config = createConfig()
+    implicit val system = createActorSystem(config)
+    implicit val materializer = ActorMaterializer()
+    implicit val executionContext = system.dispatcher
+    postStartActorSystem(system)
+
+    val host = config.getString("akka.http.server.host")
+    val port = config.getInt("akka.http.server.port")
+
+    val bindingFuture = Http().bindAndHandle(
+      handler = route,
+      interface = host,
+      port = port,
+      settings = ServerSettings(config))
+
+    bindingFuture.onComplete {
+      case Success(_) ⇒
+        println(s"Server online at http://$host:$port/")
+      case Failure(cause) ⇒
+        println(s"Error starting the server $cause")
+        preTerminateActorSystem(system)
+        system.terminate()
+    }
+
+    //It's waiting forever because maybe there is never a shutdown signal
+    Await.result(waitForShutdownSignal(system), Duration.Inf)
+
+    bindingFuture
+      .flatMap(_.unbind())
+      .onComplete(_ ⇒ {
+        preTerminateActorSystem(system)
+        system.terminate()
+      })
+  }
+
+  /**
+   * Provides the [[ActorSystem]] from where the routes will be served. Override this method, if the default [[ActorSystem]] doesn't suit your needs. You don't need to store the created instance of the [[ActorSystem]].
+   */
+  protected def createActorSystem(config: Config): ActorSystem = ActorSystem()
+
+  /**
+   * Hook that will be called once the actor system is created. Override this method if you want to do something with this [[ActorSystem]] like creating actors.
+   */
+  protected def postStartActorSystem(system: ActorSystem): Unit = {
+  }
+
+  /**
+   * Hook that will be called just before the termination of the actor system. Override this methods if you want to perform some actions on the current actorSystem, like gracefully shutting down any created actor.
+   */
+  protected def preTerminateActorSystem(system: ActorSystem): Unit = {
+  }
+
+  /**
+   * Hook that lets the user specify the future that will signal the shutdown of the server whenever completed
+   */
+  protected def waitForShutdownSignal(system: ActorSystem)(implicit ec: ExecutionContext): Future[Unit] = {
+    Future {
+      StdIn.readLine("Press RETURN to stop...\n")
+    }
+  }
```

Comment:
Yeah good API, could be implemented by external trigger completing the future... good.